File: package/website/src/types/trainTicket.ts
/**
 * src/types/trainTicket.ts
 * 火车票相关 TS 类型定义（对齐后端 Pydantic 模型）
 */

// 基础类型：铺位类型枚举（限定可选值）
export type BerthType = '上' | '中' | '下' | '无';
// 基础类型：优惠类型枚举
export type DiscountType = '学生票' | '儿童票' | '优惠票' | '全价票';
// 基础类型：座位类型枚举
export type SeatType = '一等座' | '二等座' | '商务座' | '硬座' | '软座' | '硬卧' | '软卧' | '无座' | '一等卧' | '二等卧' | '动卧' | string;

/** 火车票基础类型（创建/响应共用字段） */
export interface TrainTicketBase {
  train_code: string; // 车次号（如G1920）
  departure_station: string; // 出发站
  arrival_station: string; // 到达站
  date_time: string; // 发车时间（格式：YYYY-MM-DD HH:MM:SS，后端 datetime 转前端字符串）
  carriage: string; // 车厢号（如8A、12）
  seat_num: string; // 座位号（如12F、05下）
  berth_type: BerthType; // 铺位类型（默认"无"）
  price: number; // 票价（后端 Decimal 转前端 number）
  seat_type: SeatType; // 座位类型
  name: string; // 乘车人姓名
  discount_type: DiscountType; // 优惠类型（默认"全价票"）
  total_running_time?: number; // 总运行时间（分钟，可选，默认0）
  total_mileage?: number; // 总里程（公里，可选，默认0）
  stop_stations?: string; // 途经站点列表（可选）
  comments?: string; // 备注信息（可选）
}

/** 创建火车票请求类型（和基础类型一致，无额外字段） */
export interface TrainTicketCreate extends TrainTicketBase {}

/** 更新火车票请求类型（所有字段可选） */
export interface TrainTicketUpdate {
  train_code?: string;
  departure_station?: string;
  arrival_station?: string;
  date_time?: string;
  carriage?: string;
  seat_num?: string;
  berth_type?: BerthType;
  price?: number;
  seat_type?: SeatType;
  name?: string;
  discount_type?: DiscountType;
  total_running_time?: number;
  total_mileage?: number;
  stop_stations?: string;
  comments?: string;
}

/** 火车票响应类型（包含数据库额外字段） */
export interface TrainTicketResponse extends TrainTicketBase {
  id: number; // 数据库主键ID
  created_at: string; // 创建时间（格式：YYYY-MM-DD HH:MM:SS）
  updated_at: string; // 更新时间（格式：YYYY-MM-DD HH:MM:SS）
}

/** 火车票列表响应类型（分页） */
export interface TrainTicketListResponse {
  total: number; // 总记录数
  items: TrainTicketResponse[]; // 火车票列表数据
}

/** 获取火车票列表的查询参数类型 */
export interface TrainTicketQueryParams {
  skip?: number; // 跳过的记录数（默认0）
  limit?: number; // 每页最大记录数（默认10，最大100）
  train_code?: string; // 按车次号模糊查询
  name?: string; // 按乘车人姓名模糊查询
  departure_station?: string; // 按出发站模糊查询
  arrival_station?: string; // 按到达站模糊查询
  start_date?: string; // 发车时间起始（格式：YYYY-MM-DD）
  end_date?: string; // 发车时间结束（格式：YYYY-MM-DD）
}

File: package/server/app/schemas/train_ticket.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
@Time        : 2025/11/23 23:13
@Author      : SiYuan
@Email       : sixyuan044@gmail.com
@File        : server-train_ticket.py
@Description : 
"""
from uuid import UUID

from pydantic import BaseModel, Field
from datetime import datetime
from typing import Optional, List
from decimal import Decimal

class TrainTicketBase(BaseModel):
    """火车票基础模型（用于请求和响应）"""
    train_code: str = Field(..., description="车次号（如G1920）")
    departure_station: str = Field(..., description="出发站")
    arrival_station: str = Field(..., description="到达站")
    date_time: datetime = Field(..., description="发车时间（格式：YYYY-MM-DD HH:MM:SS）")
    carriage: str = Field(..., description="车厢号（如8A、12）")
    seat_num: str = Field(..., description="座位号（如12F、05下）")
    berth_type: Optional[str] = Field("无", description="铺位类型（上/中/下/无）")
    price: Decimal = Field(..., ge=0, description="票价（大于等于0）")
    seat_type: str = Field(..., description="座位类型（一等座/二等座/商务座等）")
    name: str = Field(..., description="乘车人姓名")
    discount_type: Optional[str] = Field("全价票", description="优惠类型（学生票/儿童票/优惠票/全价票等）")
    total_running_time: Optional[int] = Field(default=0, ge=0, description="总运行时间（分钟）")
    total_mileage: Optional[int] = Field(default=0, ge=0, description="总里程（公里）")
    stop_stations: Optional[str] = Field(None, description="途经站点列表")
    comments: Optional[str] = Field(None, description="备注信息")
    photo_id: Optional[UUID] = Field(None, description="关联照片ID")

class TrainTicketCreate(TrainTicketBase):
    """创建火车票请求模型（继承基础模型，无额外字段）"""
    pass

class TrainTicketUpdate(BaseModel):
    """更新火车票请求模型（所有字段可选）"""
    train_code: Optional[str] = Field(None, description="车次号（如G1920）")
    departure_station: Optional[str] = Field(None, description="出发站")
    arrival_station: Optional[str] = Field(None, description="到达站")
    date_time: Optional[datetime] = Field(None, description="发车时间（格式：YYYY-MM-DD HH:MM:SS）")
    carriage: Optional[str] = Field(None, description="车厢号（如8A、12）")
    seat_num: Optional[str] = Field(None, description="座位号（如12F、05下）")
    berth_type: Optional[str] = Field(None, description="铺位类型（上/中/下/无）")
    price: Optional[Decimal] = Field(None, ge=0, description="票价（大于等于0）")
    seat_type: Optional[str] = Field(None, description="座位类型（一等座/二等座/商务座等）")
    name: Optional[str] = Field(None, description="乘车人姓名")
    discount_type: Optional[str] = Field(None, description="优惠类型（学生票/儿童票/优惠票/全价票等）")
    total_running_time: Optional[int] = Field(None, description="总运行时间（分钟）")
    total_mileage: Optional[int] = Field(None, ge=0, description="总里程（公里）")
    stop_stations: Optional[str] = Field(None, description="途经站点列表")
    comments: Optional[str] = Field(None, description="备注信息")
    photo_id: Optional[UUID] = Field(None, description="关联照片ID")

class TrainTicketResponse(TrainTicketBase):
    """火车票响应模型（包含数据库额外字段）"""
    id: str
    created_at: datetime

    # 配置ORM模式（允许直接从SQLAlchemy模型转换）
    class Config:
        from_attributes = True

class TrainTicketListResponse(BaseModel):
    """火车票列表响应模型"""
    total: int = Field(..., description="总记录数")
    items: List[TrainTicketResponse] = Field(..., description="火车票列表")
